### PR TITLE
Fused HQQ Quantization Gemm

### PR DESCRIPTION
**Pull request by jeromeku**
@msaroufim 

## Fused `int4 / fp16` Quant Matmul

Fused kernel that combines asymmetric dequantization and gemm. Useful primarily for compute-bound (M > 16) scenarios and not for memory-bound / inference scenarios.

The kernel fuses two ops:

- Dequantization: upcasts `u4 / s4` weights to `float16 / bfloat16`, followed by groupwise scaling and shifting by scales / zeropoints
- GEMM: matmul on dequantized weights and activations.

Tested and benchmarked for `HQQ` but could theoretically be used for any asymmetric quantization scheme.

> **NOTE**: Benchmark below is only indicative of performance on consumer-grade `Ampere` GPUs (`A6000` specifically). When tested on `H100`, the performance is on par / marginally worse than native / compiled `torch`.  
The intended use is thus for fine-tuning / training models on non-datacenter GPUs (`80 <= compute capability < 90`). If interested in optimizing the kernel for other architectures, please drop a note in the CUDA-MODE Discord channel.

### Implementation Details

- Bitpacking is simple row interleave, no need for extensive preprocessing (e.g., `tinygemm` or `fastertransformer`)
- Tested for `float16 / bfloat16` activations, scales, and zeros
- Autotuned for both compute-bound and memory-bound configs
- Assumes operand B of the `gemm` is is the quantized type.
- Requires quantization along `in-features`, i.e., the `K` dimension, or `axis=1`, of `torch.linear.weight`.
- Implementation handles both transposed and non-tranposed quantized weights, useful for forward / backward training passes.

### Performance

Initial benchmarking (on `A6000`) demonstrates promising results, scaling well for compute-bound workloads:

|     | M    | N    | K    | group_size | dtype          | hqq_ref | triton | tinygemm |
| --- | ---- | ---- | ---- | ---------- | -------------- | ------- | ------ | -------- |
| 0   | 16   | 4096 | 4096 | 128        | torch.bfloat16 | 0.2675  | 0.0633 | 0.0382   |
| 1   | 32   | 4096 | 4096 | 128        | torch.bfloat16 | 0.2669  | 0.0704 | 0.0649   |
| 2   | 128  | 4096 | 4096 | 128        | torch.bfloat16 | 0.2689  | 0.0960 | 0.2523   |
| 3   | 256  | 4096 | 4096 | 128        | torch.bfloat16 | 0.3268  | 0.1355 | 0.5192   |
| 4   | 512  | 4096 | 4096 | 128        | torch.bfloat16 | 0.3628  | 0.2369 | 1.0892   |
| 5   | 1024 | 4096 | 4096 | 128        | torch.bfloat16 | 0.5133  | 0.4753 | 2.2016   |

- Times are in `ms`, see `benchmarks/benchmark_hqq.py`.
- `hqq_ref` is the base `HQQ_Linear` [module](https://github.com/mobiusml/hqq/blob/6d50eee4bcdd99cc10716f1297c5b2803d2b6da4/hqq/core/quantize.py#L349) that is unfused (dequantization followed by call to torch.matmul).
- `tinygemm` calls `torch.ops.aten._weight_int4pack_mm`. Implementation is a custom HQQLinear layer that wraps the preprocessing necessary for this kernel, adapted from a benchmark script posted by @mobicham from `CUDA-mode` Discord discussions.

GPU details:

```
_CudaDeviceProperties(name='NVIDIA RTX A6000', major=8, minor=6, total_memory=48676MB, multi_processor_count=84)
```

### NOTE

This implementation requires **`triton >= 3.0.0`**.

- Running tests / benchmarks requires installation of `hqq`:

  ```
  pip install hqq
  ```

### TODO
- [x] clean up / refactor test to use `pytest`
- [x] add transposed matmul
- [ ] `torch.compile` benchmarking
- [ ] introduce additional prologue / epilogue fusions
- [ ] create standalone dequant kernel
- [ ] test across range of gpus
- [ ] adapt for other quant methods (other than `hqq`)

### CODE REVIEW

**Review by msaroufim — COMMENTED** [review]
first pass

**Review by msaroufim — CHANGES_REQUESTED** [review]
first pass

File: torchao/prototype/hqq/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+| 5   | 1024 | 4096 | 4096 | 128        | torch.bfloat16 | 0.5133  | 0.4753 | 2.2016   |
+
+- Times are in `ms`, see `benchmarks/benchmark_hqq.py`.
+- `hqq_ref` is the base `HQQ_Linear` [module](https://github.com/mobiusml/hqq/blob/6d50eee4bcdd99cc10716f1297c5b2803d2b6da4/hqq/core/quantize.py#L349) that is unfused (dequantization followed by call to torch.matmul).
+- `tinygemm` calls `torch.ops.aten._weight_int4pack_mm`. Implementation is a custom HQQLinear layer that wraps the preprocessing necessary for this kernel, adapted from a benchmark script posted by @mobicham from `CUDA-mode` Discord discussions.
+
+GPU details:
+
+```
+_CudaDeviceProperties(name='NVIDIA RTX A6000', major=8, minor=6, total_memory=48676MB, multi_processor_count=84)
```
[the remark this answers is not in the document]
Comment (jeromeku) [reply torchao/prototype/hqq/README.md:32]: Attached output of pip list
[env.txt](https://github.com/pytorch-labs/ao/files/15068944/env.txt)

File: benchmarks/benchmark_hqq.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+data = []
+for shape in SHAPES:
+    for group_size in GROUP_SIZES:
+        for dtype in DTYPES:
+            timings = run_benchmark(shape, group_size, dtype)
+            data.append((*shape, group_size, dtype, *timings))
+
+
+df = pd.DataFrame(data, columns=HEADERS)
+df.to_csv("benchmark_triton.csv", index=False)
```
[the remark this answers is not in the document]
Comment (jeromeku) [reply benchmarks/benchmark_hqq.py]: fixed

**Comment by jeromeku** [comment]
@msaroufim 
- Added a `transpose` implementation for backwards passes
- Cleaned up test
  - skips unless `triton >= 3.0` + `hqq`
  - refactored to use `pytest`
  - added `transpose` test
  - removed colored outputs
- Benchmark
  - cleaned up imports
  - removed colors
  - print benchmark results to terminal rather than saving as `csv`
- Attached my `pip` env (see reply above).  TLDR: `torch 2.2.2`, `triton 3.0.0`

File: torchao/prototype/hqq/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (msaroufim) [reply torchao/prototype/hqq/README.md:32]: apologies meant `pip freeze`

[the remark this answers is not in the document]
Comment (jeromeku) [reply torchao/prototype/hqq/README.md:32]: [pip_freeze.txt](https://github.com/pytorch-labs/ao/files/15069512/pip_freeze.txt)

[the remark this answers is not in the document]
Comment (jeromeku) [reply torchao/prototype/hqq/README.md:41]: @msaroufim 

When you test on H100, can you run once with `DISABLE_MMA_V3=1`?  It toggles `Hopper` specific specializations in `triton`.  Curious to see how performance changes.

**Comment by msaroufim** [comment]
Results

(x) [marksaroufim@devvm17057.vll0 ~/ao (hqq_mixed_gemm)]$ python benchmarks/benchmark_hqq.py 
_CudaDeviceProperties(name='NVIDIA H100', major=9, minor=0, total_memory=97320MB, multi_processor_count=132)
shape=[16, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1194 Triton: 0.1678 Torch int4mm: 0.0248

shape=[32, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1203 Triton: 0.1567 Torch int4mm: 0.0442

shape=[128, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1253 Triton: 0.1607 Torch int4mm: 0.1564

shape=[256, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1269 Triton: 0.1515 Torch int4mm: 0.3025

shape=[512, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1344 Triton: 0.1477 Torch int4mm: 0.5926

shape=[1024, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1499 Triton: 0.1585 Torch int4mm: 1.1743

M,N,K,group_size,dtype,ref,triton,tinygemm
16,4096,4096,128,torch.bfloat16,0.11943938583135605,0.1677887886762619,0.024766016751527786
32,4096,4096,128,torch.bfloat16,0.12034595012664795,0.15668845176696777,0.044162582606077194
128,4096,4096,128,torch.bfloat16,0.1253334879875183,0.16072532534599304,0.15640245378017426
256,4096,4096,128,torch.bfloat16,0.1268536001443863,0.1515069156885147,0.3024669885635376
512,4096,4096,128,torch.bfloat16,0.13444344699382782,0.14771266281604767,0.5926073789596558
1024,4096,4096,128,torch.bfloat16,0.14986468851566315,0.15848150849342346,1.1743202209472656

(x) [marksaroufim@devvm17057.vll0 ~/ao (hqq_mixed_gemm)]$ DISABLE_MMA_V3=1 python benchmarks/benchmark_hqq.py 
_CudaDeviceProperties(name='NVIDIA H100', major=9, minor=0, total_memory=97320MB, multi_processor_count=132)
shape=[16, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1192 Triton: 0.1538 Torch int4mm: 0.0248

shape=[32, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1202 Triton: 0.1587 Torch int4mm: 0.0441

shape=[128, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1253 Triton: 0.1532 Torch int4mm: 0.1567

shape=[256, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1264 Triton: 0.1610 Torch int4mm: 0.3028

shape=[512, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1334 Triton: 0.1618 Torch int4mm: 0.5928

shape=[1024, 4096, 4096] group_size=128 dtype=torch.bfloat16:
Ref: 0.1496 Triton: 0.1632 Torch int4mm: 1.1754

M,N,K,group_size,dtype,ref,triton,tinygemm
16,4096,4096,128,torch.bfloat16,0.11921614408493042,0.15375947952270508,0.02484913170337677
32,4096,4096,128,torch.bfloat16,0.12021129578351974,0.15867114067077637,0.04411611706018448
128,4096,4096,128,torch.bfloat16,0.12525483965873718,0.15317335724830627,0.1566852480173111
256,4096,4096,128,torch.bfloat16,0.12637671828269958,0.16101239621639252,0.30275848507881165
512,4096,4096,128,torch.bfloat16,0.1334434598684311,0.16184988617897034,0.5928144454956055
1024,4096,4096,128,torch.bfloat16,0.14962367713451385,0.16317297518253326,1.1754025220870972

(x) [marksaroufim@devvm17057.vll0 ~/ao (hqq_mixed_gemm)]$

**Review by msaroufim — APPROVED** [review]
Thank you!
